ppa/archive/management/commands/index_pages.py

```diff
@@ -0,0 +1,106 @@
+'''
+Custom multiprocessing Solr index script for page index data.
+'''
+
+import queue
+from multiprocessing import Process, Queue, cpu_count
+from time import sleep
+
+from django.core.management.base import BaseCommand
+from parasolr.django import SolrClient, SolrQuerySet
+import progressbar
+
+from ppa.archive.models import DigitizedWork, Page
+
+
+def page_index_data(work_q, page_data_q):
+    '''Function to generate page index data and add it to
+    a queue. Takes a queue with digitized works to generate pages
+    for and a queue where page data will be added.'''
+    while True:
+        try:
+            digwork = work_q.get(True, 1)
+            # convert to the generator to a list
+            # — might be nice to chunk, but most books are small
+            # enough it doesn't matter that much
+            page_data_q.put(list(Page.page_index_data(digwork)))
+        except queue.Empty:
+            # worker is done when the work queue is empty
+            return
+
+
+def process_index_queue(index_data_q, total_to_index, work_q):
+    '''Function to send index data to Solr. Takes a
+    queue to poll for index data, a total of the items
+    to be indexed (for use with progess bar), and a work queue
+    as a way of checking that all indexing is complete.'''
+
+    solr = SolrClient()
+    progbar = progressbar.ProgressBar(redirect_stdout=True,
+                                      max_value=total_to_index)
+    count = 0
+    while True:
+        try:
+            # get data from the queue and put it into Solr
+            # block with a timeout
+            index_data = index_data_q.get(timeout=5)
+            solr.update.index(index_data)
+            # increase count based on the number of items in the list
+            count += len(index_data)
+            progbar.update(count)
+        except queue.Empty:
+            # only end if work q is also empty; otherwise, loop again
+            # indexer has just gotten ahead of page index data
+            if work_q.empty():
+                progbar.finish()
+
+                # print a summary of solr totals by item type
+                facets = SolrQuerySet().all().facet('item_type').get_facets()
+                item_totals = []
+                for item_type, total in facets.facet_fields.item_type.items():
+                    item_totals.append('%d %s%s' % (
+                        total, item_type, '' if total == 1 else 's'))
+                print('\nItems in Solr by item type: %s' %
+                      (', '.join(item_totals)))
```

Comment:
Perhaps we could take this out of this function? I feel like we might want to remove any code from Process that doesn't explicitly need multiprocessing.

Reply:
@kmcelwee That would make more sense! I tried to put it in the main handle method but couldn't get it to not run before the other work had completed, even though I ran the queue join method first. I tried a few combinations of joins to wait on the process / queue completion and couldn't get it to wait, so I just put it here. (Kind of cheating, I know).

Reply:
I got it working! I think before I was trying to join on the queue (using a `JoinableQueue`); maybe that didn't work because the page data queue can be empty temporarily when a large book is being processed. Joining on the indexer process now and it seems to be working.